### DISCUSSION TITLE
Scheduled task running like cron job?

### DISCUSSION THREAD

**Discussion by victorx98** [langgenius/dify#5175] [Suggestion]
Can we (or do we) have a time triggered mechanism like cron job? So we can schedule some of our task to run periodically at certain time? 
Thanks

**Comment by Thinking80s** [comment]
+1

**Comment by sanqianyuejia** [comment]
+1

**Comment by ZHLHZHU** [comment]
+1

**Comment by xjose97x** [comment]
For now I think one solid alternative would be to create a cron job on your computer and call the relevant Dify web API.

**Comment by y-omr** [comment]
+1

**Comment by vacheli** [comment]
+1

**Comment by 0xmalina** [comment]
+1

**Comment by fungxu** [comment]
+1

**Reply by shieldone** [reply]
+1

**Comment by yt-koike** [comment]
+1

**Comment by yt-koike** [comment]
I had a quick search and found that https://github.com/langgenius/dify/issues/19409 might be the most relevant issue to this discussion.

**Reply by yt-koike** [reply]
Or [this](https://github.com/langgenius/dify/issues/9168)?

**Comment by yt-koike** [comment]
I think I came up with a simple solution.

Let's add "Regular self-call" feature to the features (see below) of a workflow. It triggers the workflow every second with a query message saying "!cron!". You can easily check if the incoming query is a regular self-call or not by comparing the query message to that keyword.
![image](https://github.com/user-attachments/assets/61c7d68e-577b-463e-af52-54549cc4c124)


If you want a schedule feature like, triggering a workflow at 9:00 am every morning, you can use Current Time node like the workflow below.
![Cron test](https://github.com/user-attachments/assets/d431c0b7-8c66-48bb-b56a-7141c5199f77)

I think this is one of the simplest solution and is easy to implement. Still, it's a naive idea and needs reviews. For example, hackers can manualy trigger this regular self-call by setting query "!cron!". To prevent that, we can introduce a new input variable "sys.is_cron" along with "sys.query" and "sys.conversation_id". From a different perspective, self-calling every second might create too many queries to process so we could reduce the frequency to every 10 seconds or every minute then.

Any comments will help! Thank you :)

**Comment by alterxyz** [comment]
Our team is designing a trigger feature, which might include the scheduled task.

**Comment by yt-koike** [comment]
Hello again, I'm yt-koike.
I tried developing a plugin to run cron jobs.
Please feel free to use it. https://github.com/yt-koike/dify-cron
[Caution] It's still an experimental feature so be careful when you use it. I don't think it's very harmful tho. Deleting the plugin from your Dify would leave no bugs behind.
It works nicely on self-hosted servers but not so well on cloud.dify.ai.
If you have any ideas to improve this cron plugin, please make an issue or an PR to my [repo](https://github.com/yt-koike/dify-cron), thank you!

I am also planning to make a PR to publish it to marketplace. https://github.com/langgenius/dify-plugins/pull/550

**Reply by yt-koike** [reply]
It might become unnecessary after the official release of scheduled tasks.
https://github.com/langgenius/dify/discussions/5175#discussioncomment-13256149

**Reply by alterxyz** [reply]
Hi yt-koike,

Thank you for developing this plugin! Our native trigger feature will take several months to implement, so having your plugin available would be extremely valuable for the community.

I've reviewed your work and have a few important points to share about this excellent plugin:

1. **Plugin marketplace submission**: To contribute to our plugin marketplace, please submit your `.difypkg` file (the packaged single file) rather than the raw source code and individual files. You can find detailed guidance in our documentation: <https://docs.dify.ai/plugin-dev-en/0322-release-to-dify-marketplace>

2. **Cloud vs. self-hosted compatibility**: After confirming with our development team, this plugin will likely work best on self-hosted versions. Our cloud endpoints run on AWS Lambda, which may cause compatibility issues with your current implementation.

   a. **Potential improvement**: The current approach using [`started_app_ids = set()`](https://github.com/yt-koike/dify-cron/blob/5bd37013d5d3c35270353a576e0390572a8f178e/endpoints/cron.py#L9C1-L9C24) maintains state, which isn't recommended for cloud environments. Converting this to a stateless design could improve cloud compatibility.

I'll also be adding more detailed documentation about our cloud runtime architecture (hybrid AWS Lambda + Dify instance) to help developers optimize their plugins for both environments.

Thanks again for your contribution, and please let me know if you have any questions!

**Reply by yt-koike** [reply]
Hello @alterxyz 
I appreciate your comment enormously!
Sorry for accidentally putting my whole source code on [plugin submission PR](https://github.com/langgenius/dify-plugins/pull/550). I fixed it up.
Additionally, thank you for pointing out the state bit. I've been slightly concerned about that part so I want to introduce stateless design and  improve the code later on.

**Comment by yt-koike** [comment]
Hello, people.
I made a little update on Cron.
Now it works on cloud.dify.ai, supports timezones and has more detailed status.

I struggled to figure out how to make it work on cloud Dify and finally made it by integrating an API client for https://cron-job.org. I technically offloaded the calling function so cloud Dify doesn't need to keep calling itself. I kept the old code mostly untouched so it works on self-hosted servers as before.

I've already released the .difypkg on my [repo](https://github.com/yt-koike/dify-cron/releases/tag/v0.1.0) and made a PR to the [dify-plugins repo](https://github.com/langgenius/dify-plugins/pull/1162) too.

Please try it if you want to! Thank you :)

**Comment by kurokobo** [comment]
Almost there :)
https://github.com/langgenius/dify/issues/23981
